patches/api/0439-Provide-registered-listeners-in-the-PluginManager.patch

```diff
@@ -0,0 +1,215 @@
+                     i.remove();
+                     changed = true;
+                 }
+@@ -177,7 +188,11 @@ public class HandlerList {
+         boolean changed = false;
+         for (List<RegisteredListener> list : handlerslots.values()) {
+             for (ListIterator<RegisteredListener> i = list.listIterator(); i.hasNext();) {
+-                if (i.next().getListener().equals(listener)) {
++                // Paper start - faster registration
++                var j = i.next();
++                if (j.getListener().equals(listener)) {
++                    j.unregisterHandlerList(this);
++                    // Paper end
+                     i.remove();
+                     changed = true;
+                 }
+diff --git a/src/main/java/org/bukkit/plugin/PluginManager.java b/src/main/java/org/bukkit/plugin/PluginManager.java
+index 03213fde8315384ec56c16031cfc606ade2e8091..33b322078d568e174eadd424463abf745a171d12 100644
+--- a/src/main/java/org/bukkit/plugin/PluginManager.java
++++ b/src/main/java/org/bukkit/plugin/PluginManager.java
+@@ -118,6 +118,7 @@ public interface PluginManager extends io.papermc.paper.plugin.PermissionManager
+      * @param listener Listener to register
+      * @param plugin Plugin to register
+      */
++    @Deprecated(forRemoval = true) // Paper - superseded by registerListeners
+     public void registerEvents(@NotNull Listener listener, @NotNull Plugin plugin);
+ 
+     /**
+@@ -129,6 +130,7 @@ public interface PluginManager extends io.papermc.paper.plugin.PermissionManager
+      * @param executor EventExecutor to register
+      * @param plugin Plugin to register
+      */
++    @Deprecated(forRemoval = true) // Paper - superseded by registerListener
+     public void registerEvent(@NotNull Class<? extends Event> event, @NotNull Listener listener, @NotNull EventPriority priority, @NotNull EventExecutor executor, @NotNull Plugin plugin);
+ 
+     /**
+@@ -141,6 +143,7 @@ public interface PluginManager extends io.papermc.paper.plugin.PermissionManager
+      * @param plugin Plugin to register
+      * @param ignoreCancelled Whether to pass cancelled events or not
+      */
++    @Deprecated(forRemoval = true) // Paper - superseded by registerListener
+     public void registerEvent(@NotNull Class<? extends Event> event, @NotNull Listener listener, @NotNull EventPriority priority, @NotNull EventExecutor executor, @NotNull Plugin plugin, boolean ignoreCancelled);
+ 
+     /**
+@@ -316,5 +319,42 @@ public interface PluginManager extends io.papermc.paper.plugin.PermissionManager
+      */
+     @org.jetbrains.annotations.ApiStatus.Experimental
+     void overridePermissionManager(@NotNull Plugin plugin, @Nullable io.papermc.paper.plugin.PermissionManager permissionManager);
++
++    /**
++     * Registers all the events in the given listener class
++     *
++     * @param listener Listener to register
++     * @param plugin Plugin to register
++     * @return All registered listeners
```

Comment:
Clarify that this returns only registered listeners created from the supplier Listener, not *all* of them ever registered.